File: 230-kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans=-1;
    void dfs(TreeNode *root,int &k){
        if(root==NULL) return;
        dfs(root->left,k);
        if(--k==0) ans=root->val;
         dfs(root->right,k);

    }
    int kthSmallest(TreeNode* root, int k) {
        if(root==NULL) return 0;
        dfs(root,k);
        return ans;
    }
};

File: 230-kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int kthSmallest(TreeNode root, int k) {
        if(root==null) return 0;
        Stack<TreeNode>st=new Stack<>();
        while(!st.empty() || root!=null){
            while(root!=null){
                st.push(root);
                root=root.left;
            }
            TreeNode top=st.peek();
            st.pop();
            if(--k==0) return top.val;
            root=top.right;
        }
        return -1;
    }
}